crypto/zkp/primitives/commitments/damgard-fujisaki_range.go

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2017 XLAB d.o.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package commitmentzkp
+
+import (
+	"math/big"
+
+	"fmt"
+
+	"github.com/xlab-si/emmy/crypto/commitments"
+)
+
+// DFCommitmentRangeProver proves that the commitment hides a number x such that a <= x <= b.
+// Given c, prove that c = g^x * h^r (mod n) where a <= x <= b.
+type DFCommitmentRangeProver struct {
+	prover1 *DFCommitmentPositiveProver
+	prover2 *DFCommitmentPositiveProver
+}
+
+func NewDFCommitmentRangeProver(committer *commitments.DamgardFujisakiCommitter,
+	x, a, b *big.Int, challengeSpaceSize int) (*DFCommitmentRangeProver,
+	[]*big.Int, []*big.Int, []*big.Int, []*big.Int, error) {
+
+	// We will prove that b-x >= 0 and x-a >= 0.
+	bx := new(big.Int).Sub(b, x)
+	xa := new(big.Int).Sub(x, a)
+	_, r := committer.GetDecommitMsg()
+
+	rNeg := new(big.Int).Neg(r) // we act as commitment is: g^b / (g^x * h^r) = g^(b-x) * h^(-r)
+	prover1, bases1, commitmentsToSquares1, err := NewDFCommitmentPositiveProver(committer, bx, rNeg,
+		challengeSpaceSize)
+	if err != nil {
+		return nil, []*big.Int(nil), []*big.Int(nil), []*big.Int(nil), []*big.Int(nil),
+			fmt.Errorf("error in instantiating DFCommitmentPositiveProver")
+	}
+	prover2, bases2, commitmentsToSquares2, err := NewDFCommitmentPositiveProver(committer, xa, r,
+		challengeSpaceSize)
+	if err != nil {
+		return nil, []*big.Int(nil), []*big.Int(nil), []*big.Int(nil), []*big.Int(nil),
+			fmt.Errorf("error in instantiating DFCommitmentPositiveProver")
+	}
+
+	return &DFCommitmentRangeProver{
+		prover1: prover1,
+		prover2: prover2,
+	}, bases1, commitmentsToSquares1, bases2, commitmentsToSquares2, nil
+}
+
+func (p *DFCommitmentRangeProver) GetProofRandomData() []*big.Int {
+	proofRandomData1 := p.prover1.GetProofRandomData()
+	proofRandomData2 := p.prover2.GetProofRandomData()
+	return append(proofRandomData1, proofRandomData2...)
+}
+
+func (p *DFCommitmentRangeProver) GetProofData(challenges []*big.Int) []*big.Int {
+	proofData1 := p.prover1.GetProofData(challenges[0:4])
```

Comment:
L71-L72 seem dangerous because they blindly assume that `challenges` specific length. We should never make any assumptions about input - so I believe we are missing a safety check here. E.g. this code will panic if it gets a slice with only one element.

This comment actually applies to several functions (in prover as well as in verifier) in which we access input slice in such a way.